presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterStatusTracker.java

```diff
@@ -0,0 +1,146 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.router.cluster;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+import io.airlift.log.Logger;
+
+import javax.annotation.PostConstruct;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Sets.difference;
+import static com.google.common.collect.Sets.newHashSet;
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+
+public class ClusterStatusTracker
+{
+    private static final Logger log = Logger.get(ClusterStatusTracker.class);
+
+    private final ClusterManager clusterManager;
+    private final RemoteInfoFactory remoteInfoFactory;
+    private final ScheduledExecutorService queryInfoUpdateExecutor;
+
+    // Cluster status
+    private final ConcurrentHashMap<URI, RemoteClusterInfo> remoteClusterInfos = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<URI, RemoteQueryInfo> remoteQueryInfos = new ConcurrentHashMap<>();
+
+    @Inject
+    public ClusterStatusTracker(
+            ClusterManager clusterManager,
+            RemoteInfoFactory remoteInfoFactory)
+    {
+        this.clusterManager = requireNonNull(clusterManager, "clusterManager is null");
+        this.remoteInfoFactory = requireNonNull(remoteInfoFactory, "remoteInfoFactory is null");
+        this.queryInfoUpdateExecutor = newSingleThreadScheduledExecutor(threadsNamed("query-info-poller-%s"));
+    }
+
+    @PostConstruct
+    public void startPollingQueryInfo()
+    {
+        clusterManager.getAllClusters().stream()
+                .forEach(uri -> {
+                    remoteClusterInfos.put(uri, remoteInfoFactory.createRemoteClusterInfo(uri));
+                    remoteQueryInfos.put(uri, remoteInfoFactory.createRemoteQueryInfo(uri));
+                });
+
+        queryInfoUpdateExecutor.scheduleWithFixedDelay(() -> {
+            try {
+                pollQueryInfos();
+            }
+            catch (Exception e) {
+                log.error(e, "Error polling list of queries");
+            }
+        }, 5, 5, TimeUnit.SECONDS);
```

Comment:
The number of seconds works well for now at our router service. It may be a good idea to make it configurable in the config file. Users may want to config the specific interval here.

Reply:
How about extracting the numbers as a constant in this file?   Do we need to tune these numbers in the future?

Reply:
Yes, I agree. Putting them as constants in the file is good enough. We don't need to tune these numbers.
BTW, I've also checked the code in the original Presto web UI. The interval (5000 ms) there is also directly put in the file.